typings/styled-components.d.ts

```diff
@@ -1,10 +1,15 @@
 import * as React from "react";
 import { StatelessComponent, ComponentClass, PureComponent, ReactElement } from "react";
 
+import { HTMLTags, SVGTags } from "./tags";
+
 type Component<P> = ComponentClass<P> | StatelessComponent<P>;
 
+type ThemeFunction<T> = (theme: any | T) => any | T;
+type Theme<T> = any | ThemeFunction<T>;
+
 export interface ThemeProps<T> {
-  theme: T;
+  theme: Theme<T>;
```

Comment:
I think this is _inner_ props, so they should be strictly `T`, type of the theme. However, outer props can have optional `theme` as `T | (theme: T) => T`.